File: sdks/python/examples/02_payment_lifecycle.py
"""The full lifecycle: create, retrieve, authorize, capture, refund, void.

Run: PAYMENTFLOW_API_KEY=sk_test_… python examples/02_payment_lifecycle.py

Each step is one HTTP request, deliberately. This SDK has no ``create_and_capture`` helper: a
method that made two chargeable calls behind one name would leave you an authorized payment you
did not know about whenever the second one failed.
"""

from __future__ import annotations

from paymentflow import PaymentFlow


def main() -> None:
    with PaymentFlow() as client:
        # `tok_visa_approved` is a seeded test card. `client.test_helpers.list_cards()` lists
        # them all, along with what each one does.
        created = client.payments.create(
            amount_minor=5000,
            currency="USD",
            payment_method_token="tok_visa_approved",
            description="Annual subscription",
        )
        payment_id = created.get("id", "")
        print(f"created    {payment_id} {created.get('status')}")

        retrieved = client.payments.retrieve(payment_id)
        print(f"retrieved  {retrieved.get('id')} {retrieved.get('status')}")

        authorized = client.payments.authorize(payment_id)
        print(f"authorized {authorized.get('status')}, {authorized.get('amountMinor')} reserved")

        captured = client.payments.capture(payment_id)
        print(f"captured   {captured.get('capturedAmountMinor')} of {captured.get('amountMinor')}")

        # Partial refund. Omit `amount_minor` to refund everything still refundable.
        #
        # This returns the **payment**, not the refund — that is what the endpoint responds
        # with, and the new refund is in `payment["refunds"]`.
        refunded = client.payments.refund(payment_id, amount_minor=1500, reason="requested_by_customer")
        print(f"refunded   {refunded.get('refundedAmountMinor')}, status {refunded.get('status')}")
        print("refunds:   " + ", ".join(str(r.get("id")) for r in refunded.get("refunds", [])))

        # A payment that was authorized and should not be captured is voided instead, which
        # releases the reserved funds.
        to_void = client.payments.create(amount_minor=800, currency="USD")
        client.payments.authorize(to_void.get("id", ""))
        voided = client.payments.void(to_void.get("id", ""))
        print(f"voided     {voided.get('id')} {voided.get('status')}")


if __name__ == "__main__":
    main()

File: sdks/node/examples/02-payment-lifecycle.ts
/**
 * The full lifecycle: create, retrieve, authorize, capture, refund.
 *
 * Run: PAYMENTFLOW_API_KEY=sk_test_… npx tsx examples/02-payment-lifecycle.ts
 *
 * Each step is one HTTP request, deliberately. This SDK has no `createAndCapture` helper: a
 * method that made two chargeable calls behind one name would leave you an authorized payment
 * you did not know about whenever the second one failed.
 */
import { PaymentFlow } from 'paymentflow';

const client = new PaymentFlow({ apiKey: process.env['PAYMENTFLOW_API_KEY'] });

async function main(): Promise<void> {
  // `tok_visa_approved` is a seeded test card. `client.testHelpers.listCards()` lists them
  // all, along with what each one does.
  const created = await client.payments.create({
    amountMinor: 5000,
    currency: 'USD',
    paymentMethodToken: 'tok_visa_approved',
    description: 'Annual subscription',
  });
  console.log(`created   ${created.id} ${created.status}`);

  const retrieved = await client.payments.retrieve(created.id ?? '');
  console.log(`retrieved ${retrieved.id} ${retrieved.status}`);

  const authorized = await client.payments.authorize(created.id ?? '');
  console.log(`authorized ${authorized.status}, ${authorized.amountMinor} reserved`);

  const captured = await client.payments.capture(created.id ?? '');
  console.log(`captured  ${captured.capturedAmountMinor} of ${captured.amountMinor}`);

  // Partial refund. Omit `amountMinor` to refund everything still refundable.
  //
  // This returns the **payment**, not the refund — that is what the endpoint responds with,
  // and the new refund is in `payment.refunds`.
  const refunded = await client.payments.refund(created.id ?? '', {
    amountMinor: 1500,
    reason: 'requested_by_customer',
  });
  console.log(`refunded  ${refunded.refundedAmountMinor}, status ${refunded.status}`);
  console.log(`refunds:  ${(refunded.refunds ?? []).map((refund) => refund.id).join(', ')}`);

  // A payment that was authorized and should not be captured is voided instead, which
  // releases the reserved funds.
  const toVoid = await client.payments.create({ amountMinor: 800, currency: 'USD' });
  await client.payments.authorize(toVoid.id ?? '');
  const voided = await client.payments.void(toVoid.id ?? '');
  console.log(`voided    ${voided.id} ${voided.status}`);
}

main().catch((error: unknown) => {
  console.error(error);
  process.exitCode = 1;
});
